crates/ruff/src/rules/flake8_pyi/rules/fixed_return_type.rs

```diff
@@ -0,0 +1,250 @@
+/// ```
+///
+/// ## Example
+/// ```python
+/// class Foo:
+///     def __new__(cls, *args: Any, **kwargs: Any) -> Bad:
+///         ...
+///
+///     def __enter__(self) -> Bad:
+///         ...
+///
+///     async def __aenter__(self) -> Bad:
+///         ...
+///
+///     def __iadd__(self, other: Bad) -> Bad:
+///         ...
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from typing_extensions import Self
+///
+///
+/// class Foo:
+///     def __new__(cls, *args: Any, **kwargs: Any) -> Self:
+///         ...
+///
+///     def __enter__(self) -> Self:
+///         ...
+///
+///     async def __aenter__(self) -> Self:
+///         ...
+///
+///     def __iadd__(self, other: Bad) -> Self:
+///         ...
+/// ```
+/// ## References
+/// - [PEP 673](https://peps.python.org/pep-0673/)
+#[violation]
+pub struct FixedReturnType {
+    method_name: String,
+}
+
+impl Violation for FixedReturnType {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let FixedReturnType { method_name } = self;
+        format!("{method_name} methods on classes usually return `self` at runtime. Consider using `typing_extensions.Self` as return type")
+    }
+}
+
+const SELF_RETURNING_METHODS: &[&str] = &[
+    "__iadd__",
+    "__isub__",
+    "__ior__",
+    "__iand__",
+    "__imul__",
+    "__itruediv__",
+    "__ifloordiv__",
+    "__new__",
+    "__enter__",
+    "__aenter__",
+];
+
+const ITERATOR_BASES: &[&[&str]] = &[
+    &["typing", "Iterator"],
+    &["typing", "AsyncIterator"],
+    &["collections", "abc", "Iterator"],
+    &["collections", "abc", "AsyncIterator"],
+];
+
+const ASYNC_ITER_RETURN_TYPES: &[&[&str]] = &[
+    &["typing", "AsyncIterable"],
+    &["typing", "AsyncIterator"],
+    &["collections", "abc", "AsyncIterable"],
+    &["collections", "abc", "AsyncIterator"],
+];
+
+const SYNC_ITER_RETURN_TYPES: &[&[&str]] = &[
+    &["typing", "Iterable"],
+    &["typing", "Iterator"],
+    &["collections", "abc", "Iterable"],
+    &["collections", "abc", "Iterator"],
+];
+
+/// PYI034
+pub(crate) fn fixed_return_type(
+    checker: &mut Checker,
+    bases: &[Expr],
+    body: &[Stmt],
+    class_decorator_list: &[Expr],
+) {
+    // If class is final, skip
+    if class_decorator_list.iter().any(|expr| {
+        checker
+            .semantic_model()
+            .match_typing_expr(map_callable(expr), "final")
+    }) {
+        return;
+    }
+
+    let mut is_iter_subclass: bool = false;
+    if bases.len() == 1 {
+        let base = if let Expr::Subscript(ast::ExprSubscript { value, .. }) = &bases[0] {
+            // Ex) class Foo(Iterator[T]):
+            value
+        } else {
+            // Ex) class Foo(Iterator):
+            &bases[0]
+        };
+        is_iter_subclass = checker
+            .semantic_model()
+            .resolve_call_path(base)
+            .map_or(false, |call_path| {
+                ITERATOR_BASES.contains(&call_path.as_slice())
+            });
+    }
```

Comment:
Nit: I find this easier to read because I don't have to look for where the variable is mutated (I can read top down rather than having to search for the mutations)
```suggestion
    let is_iter_subclass = if bases.len() == 1 {
        let base = if let Expr::Subscript(ast::ExprSubscript { value, .. }) = &bases[0] {
            // Ex) class Foo(Iterator[T]):
            value
        } else {
            // Ex) class Foo(Iterator):
            &bases[0]
        };
        checker
            .semantic_model()
            .resolve_call_path(base)
            .map_or(false, |call_path| {
                ITERATOR_BASES.contains(&call_path.as_slice())
            });
    } else {
    	false
  	}
```

Reply:
Thanks for the review. For this one it would become:
```rust
let is_iter_subclass: bool = if class.bases.len() == 1 {
        let base = if let Expr::Subscript(ast::ExprSubscript { value, .. }) = &class.bases[0] {
            // Ex) class Foo(Iterator[T]):
            value
        } else {
            // Ex) class Foo(Iterator):
            &class.bases[0]
        };
        if checker
            .semantic_model()
            .resolve_call_path(base)
            .map_or(false, |call_path| {
                ITERATOR_BASES.contains(&call_path.as_slice())
            })
        {
            true
        } else {
            false
        }
    } else {
        false
    };
```

To cover the 3 scenarios and I find the bottom part kind of hard to read. Question of taste though so happy to go either way